### PR TITLE
Payment gateway

### PR DESCRIPTION
**Pull request by utkhagni13**

### CODE REVIEW

File: api/search.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+//Files
+const { ItemDetail } = require("../models/itemSchema");
+//SEARCH
+
+const filterSearch = (itemList, Search) => {
+  for (let wordIndex = 1; wordIndex < Search.length; wordIndex++) {
+    console.log("itemList:", itemList);
```
Comment (abhishoya) [inline api/search.js]: remove console log statements throughout the code

File: client/src/App.js [hunk 2]
```diff
@@ -27,61 +31,109 @@ const App = () => {
...
-    const URL = "/fetchItems";
-
-    const fetchJSON_fromServer = async () => {
-      const itemList = await (await fetch(URL)).json();
-      dispatch({ type: "fetchedItemJson", payload: itemList });
+    const fetchItems_fetchUser = async () => {
+      //fetching user data from the server side if any
       const userData = await (await fetch("/getUserDetails")).json();
       console.log("userData:", userData);
+      if (userData.response == true || userData.error == "not logged in") {
```
Comment (927tanmay) [inline client/src/App.js]: try to use "==="

File: client/src/App.js [hunk 3]
```diff
@@ -27,61 +31,109 @@ const App = () => {
...
+  const [loaded, setLoaded] = useState(false);
+  const [admin, setAdmin] = useState(false);
   //side effect when page first time rendered
   useEffect(() => {
-    const URL = "/fetchItems";
-
-    const fetchJSON_fromServer = async () => {
-      const itemList = await (await fetch(URL)).json();
-      dispatch({ type: "fetchedItemJson", payload: itemList });
+    const fetchItems_fetchUser = async () => {
```
Comment (abhishoya) [inline client/src/App.js]: use camel case for variable name

File: client/src/App.js [hunk 4]
```diff
@@ -27,61 +31,109 @@ const App = () => {
...
+      //fetching user data from the server side if any
       const userData = await (await fetch("/getUserDetails")).json();
```
Comment (abhishoya) [inline client/src/App.js]: don't use nested await. Either use await in different vars or use then-catch statements

File: client/src/Components/Card.js [hunk 5]
```diff
@@ -27,15 +28,16 @@ const Card = (props) => {
     const requestOptions = {
       method: "POST",
       headers: { "Content-Type": "application/json" },
-      body: JSON.stringify({ userId: user._id, itemId: id }),
+      body: JSON.stringify({ itemId: id }),
     };
     try {
-      const response = await (await fetch("/addToCart", requestOptions)).json();
-      if (response) {
+      const result = await (await fetch("/addToCart", requestOptions)).json();
```
Comment (abhishoya) [inline client/src/Components/Card.js]: don't use nested await. Either use await in different vars or use then-catch statements

File: client/src/Components/SearchBox.js [hunk 6]
```diff
@@ -1,13 +1,51 @@
-import React from "react";
+import React, { useContext, useState } from "react";
+import { Item } from "../App";
 
-const SearchBox = (props) => {
+const SearchBox = () => {
+  const [Search, setSearch] = useState("");
```
Comment (abhishoya) [inline client/src/Components/SearchBox.js:6]: s should be small

File: client/src/Components/SearchBox.js [hunk 7]
```diff
@@ -1,13 +1,51 @@
...
+  //handleSearchSubmit
+  const handleSearchSubmit = async (e) => {
+    e.preventDefault();
+    //requesting server to fetch Search data
+    const requestOptions = {
+      method: "POST",
+      headers: { "Content-Type": "application/json" },
+      body: JSON.stringify({ Search: Search }),
+    };
+    const result = await (await fetch("/search", requestOptions)).json();
```
Comment (abhishoya) [inline client/src/Components/SearchBox.js:17]: don't use nested await. Either use await in different vars or use then-catch statements

File: client/src/Components/pages/Login.js [hunk 8]
```diff
@@ -1,12 +1,15 @@
...
+import { Link } from "react-router-dom";
 
 function Login() {
   const [password, setPassword] = useState("");
   const [phone, setPhone] = useState("");
 
   return (
     <div>
-      <form className="box shadow-lg mt-5" action="/login" method="POST">
+      {/*className="login-form"*/}
```
Comment (abhishoya) [inline client/src/Components/pages/Login.js]: remove JSX comments

File: client/src/Components/pages/ShoppingCart.js [hunk 9]
```diff
@@ -54,7 +62,6 @@ function ShoppingCart({ cart, setCart }) {
       method: "POST",
       headers: { "Content-Type": "application/json" },
       body: JSON.stringify({
-        userId: user._id,
         itemId: Product.item._id,
         itemQty: Product.Qty - 1,
```
Comment (abhishoya) [inline client/src/Components/pages/ShoppingCart.js]: product.Qty can be product.qty if it's a minor change

File: client/src/Components/pages/ShoppingCart.js [hunk 10]
```diff
@@ -80,29 +90,159 @@ function ShoppingCart({ cart, setCart }) {
...
+          addOrder(
+            response.razorpay_payment_id,
+            response.razorpay_order_id,
+            response.razorpay_signature
+          );
+          alert("Your Order has been placed");
+        }
+      },
+      prefill: {
+        email: "ashes@Rajpopat.com",
```
Comment (927tanmay) [inline client/src/Components/pages/ShoppingCart.js]: Will the email and phone no. be hardcoded in prefill?

File: client/src/Components/pages/ShoppingCart.js [hunk 11]
```diff
@@ -146,12 +286,42 @@ function ShoppingCart({ cart, setCart }) {
...
           })}
+          {orderHistory ? (
+            <button className="btn btn-warning ml-3">
+              {" "}
+              <Link to="/user" className="btn text-white">
+                Order History
+              </Link>{" "}
+            </button>
+          ) : (
+            <></>
```
Comment (927tanmay) [inline client/src/Components/pages/ShoppingCart.js]: Either use null or like orderHistory && ( ...............)

File: client/src/Components/pages/UserDetails.js [hunk 12]
```diff
@@ -1,52 +1,235 @@
...
-        </div>
-        <br />
-        <div className="form-group">
-          <h5>Email address</h5>
-          <h6 style={{ color: "blue" }} onClick={() => {}}>
-            Edit
-          </h6>
-          <input type="text" value={user.email} name="Email" disabled />
+    <>
+      <div className="product flex-container userbackground">
```
Comment (abhishoya) [inline client/src/Components/pages/UserDetails.js]: userbackground can be like user-background. Try follow snake case for CSS classes

File: client/src/Components/pages/UserDetails.js [hunk 13]
```diff
@@ -1,52 +1,235 @@
...
+                          .substring(
+                            0,
+                            element.dateOfOrder.toString().length - 30
+                          )}
+                      </p>
+                      <p>
+                        <b>Order ID : </b>
+                        {element._id}
+                      </p>
+                      {element.order.map((item, i) => {
```
Comment (abhishoya) [inline client/src/Components/pages/UserDetails.js:114]: item doesn't make sense here. can use a more semantic variable name

File: models/userScema.js [hunk 14]
```diff
@@ -29,6 +31,7 @@ const userDetailSchema = mongoose.Schema({
   password: {
     type: String,
     required: true,
+    unique: true,
```
Comment (abhishoya) [inline models/userScema.js:34]: why is password unique true?

**Review by abhishoya — CHANGES_REQUESTED** [review]
The comments maybe at only one place, but if you see a similar mistake somewhere else too, do rectify it.

**Pull request merged**